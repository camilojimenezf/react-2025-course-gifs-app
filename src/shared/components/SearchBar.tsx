import {
  useEffect,
  useState,
  type ChangeEvent,
  type KeyboardEvent,
} from "react";

interface SearchBarProps {
  placeholder?: string;
  buttonLabel?: string;

  onQuery: (term: string) => void;
}

export const SearchBar = ({
  placeholder,
  buttonLabel,
  onQuery,
}: SearchBarProps) => {
  const [query, setQuery] = useState("");

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      onQuery(query);
    }, 700);

    return () => {
      clearTimeout(timeoutId);
    };
  }, [query, onQuery]);

  const handleSearch = () => {
    onQuery(query);
  };

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    setQuery(event.target.value);
  };

  const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter") {
      handleSearch();
    }
  };

  return (
    <div className="search-container">
      <input
        type="text"
        placeholder={placeholder ?? ""}
        value={query}
        onChange={handleOnChange}
        onKeyDown={handleKeyDown}
      />
      <button onClick={handleSearch}>{buttonLabel ?? "Buscar"}</button>
    </div>
  );
};
