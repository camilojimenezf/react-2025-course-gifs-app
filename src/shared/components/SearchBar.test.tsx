import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { describe, expect, test, vi } from "vitest";
import { SearchBar } from "./SearchBar";

describe("SearchBar", () => {
  test("should render the component correctly", () => {
    const { container } = render(<SearchBar onQuery={() => {}} />);

    expect(container).toMatchSnapshot();
  });

  test("should call onQuery with delay (debounce)", async () => {
    const onQueryMock = vi.fn();
    const inputText = "test message";
    render(<SearchBar onQuery={onQueryMock} />);

    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: inputText },
    });

    expect(onQueryMock).not.toHaveBeenCalled(); // not called immediately

    await new Promise((resolve) => setTimeout(() => resolve(true), 200));

    expect(onQueryMock).not.toHaveBeenCalled(); // not called in 200 ms

    // should be called after 200 ms
    await waitFor(() => {
      expect(onQueryMock).toHaveBeenCalled();
      expect(onQueryMock).toHaveBeenCalledWith(inputText);
    });
  });

  test("should call onQuery with the latest change (debounce)", async () => {
    const onQueryMock = vi.fn();
    const inputText = "test message";
    render(<SearchBar onQuery={onQueryMock} />);

    const input = screen.getByRole("textbox");

    fireEvent.change(input, {
      target: { value: "t" },
    });
    fireEvent.change(input, {
      target: { value: "te" },
    });
    fireEvent.change(input, {
      target: { value: "tes" },
    });
    fireEvent.change(input, {
      target: { value: "test" },
    });
    fireEvent.change(input, {
      target: { value: "test m" },
    });
    fireEvent.change(input, {
      target: { value: inputText },
    });

    await waitFor(() => {
      expect(onQueryMock).toHaveBeenCalled();
      expect(onQueryMock).toHaveBeenCalledWith(inputText);
    });
  });

  test("should call immediately onQuery when clicking button", () => {
    const onQueryMock = vi.fn();
    const inputText = "test message";
    render(<SearchBar onQuery={onQueryMock} />);

    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    fireEvent.change(input, {
      target: { value: inputText },
    });

    fireEvent.click(button);

    expect(onQueryMock).toHaveBeenCalled();
    expect(onQueryMock).toHaveBeenCalledWith(inputText);
  });

  test("should use the placeholder", () => {
    const placeholder = "insert message";

    render(<SearchBar onQuery={() => {}} placeholder={placeholder} />);

    const input = screen.getByRole("textbox");
    const placeholderElement = screen.getByPlaceholderText(placeholder);

    expect(input).toBe(placeholderElement);
  });
});
