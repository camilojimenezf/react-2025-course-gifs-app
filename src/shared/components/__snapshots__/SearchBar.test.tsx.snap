// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar > should render the component correctly 1`] = `
<div>
  <div
    class="search-container"
  >
    <input
      placeholder=""
      type="text"
      value=""
    />
    <button>
      Buscar
    </button>
  </div>
</div>
`;
